### PR TITLE
[WIP] Optimize queries and hydration of record navigation properties

### PR DESCRIPTION
**Pull request by aztech-dev**
- Remove n+1 selects on a lot of entities (querying records should have a fixed number of queries now)

### CODE REVIEW

File: lib/classes/caption/Field/Value.php [hunk 1]
```diff
@@ -19,6 +19,85 @@
...
+        if (empty($ids)) {
+            return array();
+        }
+
+        $connection = $record->get_databox()->get_connection();
+
+        $query = 'SELECT id, record_id, meta_struct_id, value, VocabularyType, VocabularyId
+            FROM metadatas WHERE id IN (%s)';
+
+        $query = sprintf($query, implode(', ', $ids));
```
Comment (bburnichon) [inline lib/classes/caption/Field/Value.php]: SQL injection possible. Can you please use DBAL execute instead?

File: lib/classes/caption/Field/Value.php [hunk 2]
```diff
@@ -73,93 +152,72 @@ class caption_Field_Value implements cache_cacheableInterface
...
-            $this->VocabularyId = $row['VocabularyId'];
+            $value->VocabularyType = $row['VocabularyType'] ? Vocabulary\Controller::get($value->app, $row['VocabularyType']) : null;
+            $value->VocabularyId = $row['VocabularyId'];
         } catch (\InvalidArgumentException $e) {
 
         }
 
-        if ($this->VocabularyType) {
+        if ($value->VocabularyType) {
             /**
```
Comment (bburnichon) [inline lib/classes/caption/Field/Value.php]: Can you also change phpdocs into simple comments?

File: lib/classes/record/adapter.php [hunk 3]
```diff
@@ -31,18 +31,251 @@
...
+        }
+
+        $query = 'SELECT r.coll_id, r.record_id, r.credate , r.uuid, r.moddate, r.parent_record_id,
+                    r.type, r.originalname, r.bitly, r.sha256, r.mime, BIN(r.status) AS status,
+                    s.subdef_id, s.name, s.file, s.width, s.height, s.mime,
+                    s.path, s.size, s.substit, s.created_on, s.updated_on, s.etag
+                  FROM record AS r LEFT JOIN subdef AS s ON (s.record_id = r.record_id)
+                  WHERE r.record_id IN (%s) ORDER BY s.record_id, s.subdef_id';
+
+        $query = sprintf($query, implode(', ', $ids));
```
Comment (bburnichon) [inline lib/classes/record/adapter.php]: again, SQL injection possible

File: lib/classes/databox/field.php [hunk 4]
```diff
@@ -28,88 +28,227 @@
...
+    public static function get_all(Application $app, databox $databox)
+    {
+        $connection = $databox->get_connection();
+      
+        $query = "SELECT `id`, `thumbtitle`, `separator`, `dces_element`, `tbranch`,
+                         `type`, `report`, `multi`, `required`, `readonly`, `indexable`,
+                         `name`, `src`, `business`, `VocabularyControlType`,
+                         `RestrictToVocabularyControl`, `sorter`,
+                         `label_en`, `label_fr`, `label_de`, `label_nl`
+                  FROM metadatas_structure";
```
Comment (aztech-dev) [inline lib/classes/databox/field.php]: Add SORT statement

File: lib/classes/media/subdef.php [hunk 5]
```diff
@@ -23,7 +23,181 @@
...
+        return new self($app, $record, $name, $substitute, true);
+    }
+
+    public static function getMany(Application $app, record_adapter $record, array $names)
+    {
+        $connection = $record->get_databox()->get_connection();
+        $query = 'SELECT s.subdef_id, s.name, s.file, s.width, s.height, s.mime,
+                         s.path, s.size, s.substit, s.created_on, s.updated_on, s.etag
+                  FROM subdef s
+                  WHERE s.name IN (%s) AND s.record_id = :record_id';
```
Comment (aztech-dev) [inline lib/classes/media/subdef.php]: Check if sort is necessary

File: lib/classes/caption/Field/Value.php [hunk 6]
```diff
@@ -19,6 +19,91 @@
...
+    public static function getMany(Application $app, record_adapter $record, array $ids)
+    {
+        if (empty($ids)) {
+            return array();
+        }
+
+        $connection = $record->get_databox()->get_connection();
+
+        $query = 'SELECT id, record_id, meta_struct_id, value, VocabularyType, VocabularyId
+            FROM metadatas WHERE id IN (%s)';
```
Comment (aztech-dev) [inline lib/classes/caption/Field/Value.php]: Check sorting

File: lib/classes/caption/record.php [hunk 7]
```diff
@@ -22,8 +22,120 @@
...
+     * @param record_adapter[] $records
+     * @return caption_record[]
+     */
+    public static function getMany(Application $app, databox $databox, array $records)
+    {
+        $query = "SELECT m.record_id, m.id as meta_id, m.id as id, s.id as structure_id,
+                         m.value, m.VocabularyType, m.VocabularyId
+                  FROM metadatas m INNER JOIN metadatas_structure s ON (s.id = m.meta_struct_id)
+                  WHERE m.record_id IN (%s)
+                  ORDER BY m.record_id, s.id, s.sorter ASC";
```
Comment (aztech-dev) [inline lib/classes/caption/record.php]: Check sort by s.id

File: lib/classes/databox.php [hunk 8]
```diff
@@ -16,6 +16,79 @@
 
 class databox extends base
 {
+
+    public static function getAll(Application $app, \appbox $appbox)
+    {
+        try {
+            $rows = $appbox->get_data_from_cache(static::CACHE_BASE_DATABOX . '_all');
```
Comment (bburnichon) [inline lib/classes/databox.php]: Be careful, no exception is thrown by memcache when value is not present. it sometimes only return `false` instead.

File: lib/classes/record/adapter.php [hunk 9]
```diff
@@ -31,18 +31,257 @@
...
+                return (int) $id['id'];
+            }
+
+            return (int) $id;
+        }, $record_ids);
+
+        $numbers = array();
+
+        foreach ($record_ids as $id) {
+            $numbers[$id['id']] = null;
```
Comment (bburnichon) [inline lib/classes/record/adapter.php]: use of `$id` as an array. You cannot be sure of this. You did normalize `$ids` for this, didn't you?

**Comment by aztech-dev** [comment]
Closing this P/R as it targets 3.8

**Pull request closed without merging**